### PR TITLE
Fix for #2543 -  XpressionMapper does not convert the delegate return…

### PR DESCRIPTION
**Pull request by BlaiseD**
… type.

Fixes issue #2543 .

### CODE REVIEW

**Comment by BlaiseD** [comment]
@HRasch - the tests in #2542 seem to work for both expression (Xpression) mapping and UseAsDatSource if the configuration is modified. [See CustomStructTests](https://github.com/BlaiseD/AutoMapper/blob/master/src/UnitTests/CustomStructTests.cs). Sounds like you would like to make this work by specifying the types alone.  That would be great.  Note #2378 for any performance impact.

File: src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs [hunk 1]
```diff
@@ -33,8 +34,18 @@ public static TDestDelegate MapExpression<TDestDelegate>(this IMapper mapper, La
             var remappedBody = visitor.Visit(expression.Body);
             if (remappedBody == null)
                 throw new InvalidOperationException(Resource.cantRemapExpression);
+
+            remappedBody = VerifyType(typeDestFunc.GetGenericArguments().Last(), remappedBody);
```
Comment (lbargaoanu) [inline src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs]: Simply replacing VerifyType with ToType should be enough here.

Comment (BlaiseD) [reply src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs]: Good  point.

File: src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs [hunk 2]
```diff
@@ -33,9 +34,15 @@ public static TDestDelegate MapExpression<TDestDelegate>(this IMapper mapper, La
...
             if (remappedBody == null)
                 throw new InvalidOperationException(Resource.cantRemapExpression);
+
+            return (TDestDelegate)Lambda
+                                    (
+                                        typeDestFunc,
+                                        ExpressionFactory.ToType(remappedBody, typeDestFunc.GetGenericArguments().Last()),
+                                        expression.GetDestinationParameterExpressions(visitor.InfoDictionary, typeMappings)
+                                    );
+        }
```
Comment (lbargaoanu) [inline src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs]: The two MapExpression methods are almost the same, only the visitor differs. Some code reuse seems in order.

**Comment by lbargaoanu** [comment]
That convert code seems ad hoc. Because we already have that code in the mappers, so maybe it should try to use those as the normal mapping does?

File: src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs (hunk 2 above)
Comment (BlaiseD) [reply src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs]: I agree. The simplest method would be to pass a `Func<Dictionary<Type, Type>, XpressionMapperVisitor> func` to the consolidated method.  That belongs in a separate PR though - we did not add the two methods in this PR -  unless you want to fix it.

**Comment by BlaiseD** [comment]
Please be specific about how you would do it differently.

File: src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs (hunk 2 above)
Comment (lbargaoanu) [reply src/AutoMapper/XpressionMapper/Extensions/MapperExtensions.cs]: Nah, I'm good :)

**Comment by lbargaoanu** [comment]
I was talking about the convert code in that other PR. I don't really know what's all that about. But if the types don't match, I was thinking maybe mapping to the needed type would help. As we already have mappers that consider the conversion operators that might apply. Does that make sense? :)

**Comment by BlaiseD** [comment]
Ah - sorry I misread you.  It's possible. I'll have to look at the code further.  The challenge is with binary expressions e.g. int? == int, CustomString == System.String etc...  Any side could be a method call, member, variable, or constant. The base ExpressionVisitor.VisitBinary will throw an exception when the two sides do not match (unless one side is converted in the override).  From #2378 - the fix could be expensive - or maybe it is easier than I thought.

**Comment by lbargaoanu** [comment]
Then I guess mapping doesn't really make sense.

**Comment by HRasch** [comment]
Excuse me if I'm wrong, but the compare on the VisitBinary is just one case. OData e.g. tries to create projections with MemberInit and if the types are not equal, exceptions are thrown. I tried to fix the problem by introducing a new parameter to the MappingVisitor-Class, telling what return type is expected, so I can decide if i have to convert the member to a specific type (by a typemap-lookup or implicit cast). But i failed because of calls like _parentMappingVisitor.Visit, where I have no control over the visitor.
In the end it feelt like the MappingVisitor is just doing to much and maybe another visitor should be introduced, just for the type-convertings. Then, in case of a binary expression, the left expression should control the expected type of the right expression. This should also work with MemberInit and MemberBind.

**Comment by BlaiseD** [comment]
@jbogard @lbargaoanu - Do you have reservations about this one? Seems like it's been a while.

**Comment by lbargaoanu** [comment]
I could go either way :), so I just left it for @jbogard.

**Comment by BlaiseD** [comment]
Clear - I think. There are related changes I'd like to make so I'm going to close this one and address #2543 in a different PR.  Thanks.

**Pull request closed without merging**